### PR TITLE
Torch based SAC

### PR DESCRIPTION
**Pull request by ShuaibinLi**
### Add files:
mujoco_agent.py, mujoco_model.py, train.py, README.md, ---> benchmark/torch/
sac.py ---> parl/algorithms/torch/
### Change file:
parl/algorithms/torch/init.py

### CODE REVIEW

File: benchmark/torch/sac/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+> Paper: SAC in [Soft Actor-Critic: Off-Policy Maximum Entropy Deep Reinforcement Learning with a Stochastic Actor](https://arxiv.org/abs/1801.01290)
+
+### Mujoco games introduction
+Please see [here](https://github.com/openai/mujoco-py) to know more about Mujoco games.
+
+### Benchmark result
+
+<img src=".benchmark/SAC_results.png" width = "800" height ="400" alt="SAC_HalfCheetah-v1"/>
+
++ We trained our model in a Mujoco environment: "HalfCheetah-v1".
```
Comment (rical730) [inline benchmark/torch/sac/README.md]: please remove this line

Comment (rical730) [inline benchmark/torch/sac/README.md]: please remove this line because environment name has been mentioned in picture

File: benchmark/torch/sac/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
++ We trained the model 3 times(for 3 different seeds).
```
Comment (rical730) [inline benchmark/torch/sac/README.md]: Each experiment was run three times with different seeds

File: benchmark/torch/sac/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+--alpha, default=0.2              # Temperature parameter, determines the relative importance of entropy term against the reward
+--batch_size, default=256         # Batch size for learning
+--discount, default=0.99          # Discount factor
+--tau, default=5e-3               # Target network update rate
+--actor_lr, default=3e-4          # Learning rate of actor network
+--critic_lr, default=3e-4         # Learning rate of critic network
+--policy_freq, default=1          # Frequency to train actor and update params
+--automatic_entropy_tuning, default=False  # Automatically adjust α
+--entropy_lr, default=3e-4        # Learning rate of entropy
+```
```
Comment (rical730) [inline benchmark/torch/sac/README.md]: please remove arguments' doc here, show them in code

File: benchmark/torch/sac/train.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            agent.learn(batch_obs, batch_action, batch_reward, batch_next_obs,
+                        batch_terminal)
+
+    return episode_reward, episode_steps
+
+
+# Runs policy for 5 episodes by default and returns average reward
+# A fixed seed is used for the eval environment
+def run_evaluate_episodes(agent, env_name, seed, eval_episodes=5):
+    eval_env = gym.make(env_name)
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please make eval_env in main()

File: parl/algorithms/torch/sac.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+import parl
+import torch
+from torch.distributions import Normal
+import torch.nn.functional as F
+from copy import deepcopy
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+# To enforcing Action Bound
+epsilon = 1e-6
```
Comment (rical730) [inline parl/algorithms/torch/sac.py]: please remove epsilon here

File: parl/algorithms/torch/sac.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        act_mean, act_log_std = self.model.policy(obs)
+        normal = Normal(act_mean, act_log_std.exp())
+        # for reparameterization trick  (mean + std*N(0,1))
+        x_t = normal.rsample()
+        y_t = torch.tanh(x_t)
+        action = y_t * self.max_action
+
+        log_prob = normal.log_prob(x_t)
+        # Enforcing Action Bound
+        log_prob -= torch.log(self.max_action * (1 - y_t.pow(2)) + epsilon)
```
Comment (rical730) [inline parl/algorithms/torch/sac.py]: epsilon --> 1e-6

File: parl/algorithms/torch/sac.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+        self.model = model.to(device)
+        self.target_model = deepcopy(self.model)
+        self.actor_optimizer = torch.optim.Adam(
+            self.model.get_actor_params(), lr=actor_lr)
+        self.critic_optimizer = torch.optim.Adam(
+            self.model.get_critic_params(), lr=critic_lr)
+
+        if self.automatic_entropy_tuning is True:
+            # target entropy = - dim(action_space) as given in the paper (e.g. -6 for HalfCheetah)
```
Comment (rical730) [inline parl/algorithms/torch/sac.py]: please remove this comment line

File: benchmark/torch/sac/README.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+#### Train
+```
+# To train an agent for HalfCheetah-v1 game
+python train.py
+
+# To train for other game & for automatic entropy tuning
+# python train.py --env [ENV_NAME] ---automatic_entropy_tuning True
```
Comment (rical730) [inline benchmark/torch/sac/README.md]: --automatic_entropy_tuning

Comment (ShuaibinLi) [reply benchmark/torch/sac/README.md]: Resolved.

File: benchmark/torch/sac/train.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            agent.learn(batch_obs, batch_action, batch_reward, batch_next_obs,
+                        batch_terminal)
+
+    return episode_reward, episode_steps
+
+
+# Runs policy for 5 episodes by default and returns average reward
+# A fixed seed is used for the eval environment
+def run_evaluate_episodes(agent, eval_env, eval_episodes=5):
+
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please remove this blank line

File: benchmark/torch/sac/train.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+from parl.algorithms import SAC
+
+
+# Run episode for training
+def run_train_episode(agent, env, rpm):
+
+    obs, done = env.reset(), False
+    episode_reward, episode_steps = 0, 0
+    while not done:
+
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please remove this blank line

File: benchmark/torch/sac/train.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+import argparse
+from parl.utils import logger, tensorboard, ReplayMemory
+from mujoco_model import MujocoModel
+from mujoco_agent import MujocoAgent
+from parl.algorithms import SAC
+
+
+# Run episode for training
+def run_train_episode(agent, env, rpm):
+
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please remove this blank line

File: benchmark/torch/sac/train.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    agent = MujocoAgent(algorithm, state_dim, action_dim)
+    rpm = ReplayMemory(
+        max_size=int(1e6), obs_dim=state_dim, act_dim=action_dim)
+
+    episode_num = 0
+    total_steps = 0
+    test_flag = 0
+
+    while total_steps < args.max_timesteps:
+
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please remove this blank line

File: benchmark/torch/sac/train.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+            tensorboard.add_scalar('eval/episode_reward', avg_reward,
+                                   total_steps)
+            logger.info('Evaluation over: {} episodes, Reward: {}'.format(
+                args.eval_episodes, avg_reward))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--env", default="HalfCheetah-v1")  # Mujoco gym environment name
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please move comment into `help` argument

File: parl/algorithms/torch/sac.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+__all__ = ['SAC']
+
+
+class SAC(parl.Algorithm):
+    def __init__(self,
+                 model,
+                 max_action,
```
Comment (rical730) [inline parl/algorithms/torch/sac.py]: why should we need to pass `max_action` while we have `action_space` in parameter list ?

File: parl/algorithms/torch/sac.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        action = y_t * self.max_action
+
+        log_prob = normal.log_prob(x_t)
+        # Enforcing Action Bound
+        log_prob -= torch.log(self.max_action * (1 - y_t.pow(2)) + 1e-6)
+        log_prob = log_prob.sum(1, keepdims=True)
+        return action, log_prob
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+
```
Comment (rical730) [inline parl/algorithms/torch/sac.py:84]: please remove this blank line here

File: parl/algorithms/torch/sac.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        y_t = torch.tanh(x_t)
+        action = y_t * self.max_action
+
+        log_prob = normal.log_prob(x_t)
+        # Enforcing Action Bound
+        log_prob -= torch.log(self.max_action * (1 - y_t.pow(2)) + 1e-6)
+        log_prob = log_prob.sum(1, keepdims=True)
+        return action, log_prob
+
+    def learn(self, obs, action, reward, next_obs, terminal):
```
Comment (rical730) [inline parl/algorithms/torch/sac.py:80]: please split `learn()` function to `_actor_learn()` and `_critic_learn()`, it will be more readable.

File: parl/algorithms/torch/sac.py (hunk 14 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/sac.py]: The action must be distributed between action_space.low[0] and action_space.high[0], which are opposite numbers(generally it's the case in gym envs).
So in this case, we can let 
```
max_action = action_space.high[0]
action =  torch.tanh(input) * max_action
```

File: parl/algorithms/torch/sac.py (hunk 16 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/sac.py:80]: done

File: parl/algorithms/torch/sac.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+                automatic_entropy_tuning(bool): whether or not to adjust alpha automatically
+                entropy_lr: learning rate of entropy
+                action_dim: the dimension of an action, env.action_space.shape[0], to calculate target_entropy
+        """
+        assert isinstance(discount, float)
+        assert isinstance(tau, float)
+        assert isinstance(alpha, float)
+        assert isinstance(actor_lr, float)
+        assert isinstance(critic_lr, float)
+        assert isinstance(entropy_lr, float)
```
Comment (rical730) [inline parl/algorithms/torch/sac.py]: Is `entropy_lr` a necessary parameter?  if not, please remove this assert line.

Comment (ShuaibinLi) [reply parl/algorithms/torch/sac.py]: done

File: benchmark/torch/sac/train.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+from parl.algorithms import SAC
+
+WARMUP_STEPS = 1e4
+EVAL_EVERY_STEPS = 5e3
+EVAL_EPISODES = 5
+MEMORY_SIZE = int(1e6)
+BATCH_SIZE = 256
+GAMMA = 0.99
+TAU = 0.005
+ALPHA = 0.2  # determines the relative importance of entropy term against the reward
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/sac/train.py]: Put it at the parser. The various benchmarks require different values for this argument.

Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/sac/train.py]: Please notify the user at the readme that Humanoid requires 0.2 while other environments require 0.05 .

Comment (ShuaibinLi) [reply benchmark/torch/sac/train.py]: done

File: benchmark/torch/sac/README.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
++ mujoco-py==0.5.7
+
+### Start Training:
+#### Train
+```
+# To train for HalfCheetah-v1(default),Hopper-v1,Walker2d-v1,Ant-v1
+# --alpha 0.2(default)
+python train.py --env [ENV_NAME]
+
+# To train for Humanoid-v1
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/sac/README.md]: To reproduce the performance of Humanoid-v1

File: benchmark/torch/sac/train.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--env", default="HalfCheetah-v1", help='Mujoco gym environment name')
+    parser.add_argument(
+        "--seed",
+        default=0,
+        type=int,
+        help='Sets Gym, PyTorch and Numpy seeds')
+    parser.add_argument(
+        "--max_timesteps",
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please replace `max_timesteps` by `train_total_steps`, and add `test_every_steps` in add_argument, refer https://github.com/PaddlePaddle/PARL/blob/develop/examples/SAC/train.py#L139

File: benchmark/torch/sac/mujoco_model.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
+import parl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+# clamp bounds for Std of action_log
+LOG_SIG_MAX = 2.0
+LOG_SIG_MIN = -20.0
+
```
Comment (rical730) [inline benchmark/torch/sac/mujoco_model.py:23]: please Initialize model weights

```
def weights_init_(m):
    if isinstance(m, nn.Linear):
        torch.nn.init.xavier_uniform_(m.weight, gain=1)
        torch.nn.init.constant_(m.bias, 0)
```

File: parl/algorithms/torch/sac.py [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+        # Enforcing Action Bound
+        log_prob -= torch.log((1 - action.pow(2)) + 1e-6)
+        log_prob = log_prob.sum(1, keepdims=True)
+        return action, log_prob
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+        self._critic_learn(obs, action, reward, next_obs, terminal)
+        self._actor_learn(obs)
+
+        self.sync_target()
```
Comment (rical730) [inline parl/algorithms/torch/sac.py:85]: return critic_cost, actor_cost

File: parl/algorithms/torch/sac.py [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+            target_Q = torch.min(q1_next, q2_next) - self.alpha * next_log_pro
+            target_Q = reward + self.gamma * terminal * target_Q
+        cur_q1, cur_q2 = self.model.critic_model(obs, action)
+
+        critic_loss = F.mse_loss(cur_q1, target_Q) + F.mse_loss(
+            cur_q2, target_Q)
+
+        self.critic_optimizer.zero_grad()
+        critic_loss.backward()
+        self.critic_optimizer.step()
```
Comment (rical730) [inline parl/algorithms/torch/sac.py:102]: return critic_loss

File: parl/algorithms/torch/sac.py [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+
+    def _actor_learn(self, obs):
+        act, log_pi = self.sample(obs)
+        q1_pi, q2_pi = self.model.critic_model(obs, act)
+        min_q_pi = torch.min(q1_pi, q2_pi)
+        actor_loss = ((self.alpha * log_pi) - min_q_pi).mean()
+
+        self.actor_optimizer.zero_grad()
+        actor_loss.backward()
+        self.actor_optimizer.step()
```
Comment (rical730) [inline parl/algorithms/torch/sac.py:113]: return actor_loss

File: parl/algorithms/torch/sac.py [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+        with torch.no_grad():
+            next_action, next_log_pro = self.sample(next_obs)
+            q1_next, q2_next = self.target_model.critic_model(
+                next_obs, next_action)
+            target_Q = torch.min(q1_next, q2_next) - self.alpha * next_log_pro
+            target_Q = reward + self.gamma * terminal * target_Q
+        cur_q1, cur_q2 = self.model.critic_model(obs, action)
+
+        critic_loss = F.mse_loss(cur_q1, target_Q) + F.mse_loss(
+            cur_q2, target_Q)
```
Comment (rical730) [inline parl/algorithms/torch/sac.py:96]: critic_loss = critic_loss.mean()

File: parl/algorithms/torch/sac.py [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+        self.sync_target()
+
+    def _critic_learn(self, obs, action, reward, next_obs, terminal):
+        with torch.no_grad():
+            next_action, next_log_pro = self.sample(next_obs)
+            q1_next, q2_next = self.target_model.critic_model(
+                next_obs, next_action)
+            target_Q = torch.min(q1_next, q2_next) - self.alpha * next_log_pro
+            target_Q = reward + self.gamma * terminal * target_Q
```
Comment (rical730) [inline parl/algorithms/torch/sac.py]: (1.0 - terminal)

File: benchmark/torch/sac/train.py [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+        # Select action randomly or according to policy
+        if rpm.size() < WARMUP_STEPS:
+            action = np.random.uniform(-1, 1, size=action_dim)
+        else:
+            action = agent.sample(obs)
+
+        # Perform action
+        next_obs, reward, done, _ = env.step(action)
+        terminal = float(done) if episode_steps < env._max_episode_steps else 0
+        terminal = 1. - terminal
```
Comment (rical730) [inline benchmark/torch/sac/train.py]: please remove this line. terminal=False before episode finished

File: benchmark/torch/sac/mujoco_agent.py [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+        super(MujocoAgent, self).__init__(algorithm)
+
+        self.device = torch.device("cuda" if torch.cuda.
+                                   is_available() else "cpu")
+
+        self.alg.sync_target(decay=0)
+
+    def predict(self, obs):
+        obs = torch.FloatTensor(obs.reshape(1, -1)).to(self.device)
+        action, _ = self.alg.predict(obs)
```
Comment (rical730) [inline benchmark/torch/sac/mujoco_agent.py]: `action = self.alg.predict(obs)`

`self.alg.predict` returns one value

**Review by rical730 — APPROVED** [review]
LGTM~
